reana_server/utils.py

```diff
@@ -90,7 +90,7 @@ def get_usage_percentage(usage, limit):
 def publish_workflow_submission(workflow, user_id, total_cluster_memory, parameters):
     """Publish workflow submission."""
     complexity = _calculate_complexity(workflow)
```

Comment:
I guess we don't even need to run `_calculate_complexity` if we're in `fifo` mode but just set the priority to 0? It'd improve the performance. 

Just watch out because then we would need to move the `RunStatus.queued` assignment outside. I think we can just use `Workflow.update_workflow_status`

Reply:
> I guess we don't even need to run `_calculate_complexity` if we're in `fifo` mode but just set the priority to 0? It'd improve the performance.

Yes, that's why I've added this peace of code to `estimate_complexity()` before any calculations:
```python
# No need to estimate the complexity for "fifo" strategy
if scheduling_policy == "fifo":
    return []
```

> It's a bit strange that the logic to decide whether we should calculate complexity is inside a function called `_calculate_complexity`/`estimate_complexity`.

I was thinking where to put the policy check and for me it looked cleaner to keep this logic inside `estimate_complexity()` rather than in `utils.py`, but it could be done either way I guess

Reply:
IMO, if the policy is `fifo`, there shouldn't be any complexity estimation function involved.

I understand your point, it's just a matter of keeping things separated and more legible. I'd like to have more points of view on this. cc/ @tiborsimko

Reply:
Refactor done [here](https://github.com/reanahub/reana-server/pull/377/commits/a029855d14d42ca187236bbfffcb4406ff27f003).

Reply:
Tested the update, and I saw `r-server` scheduler pod issues:

```
$ kubectl logs reana-server-694d7955ff-7gck7 scheduler
...
  File "/usr/local/lib/python3.8/site-packages/kombu/messaging.py", line 590, in <listcomp>
    [callback(body, message) for callback in callbacks]
  File "/usr/local/lib/python3.8/site-packages/reana_server/scheduler.py", line 236, in on_message
    sleep(REANA_SCHEDULER_SECONDS_TO_WAIT_FOR_REANA_READY)
TypeError: an integer is required (got type str)
```

which "self-healed" after pod restarted a few times...

At least I've got to test the how the web ui looks like :wink: 

![reana-service-unavailable](https://user-images.githubusercontent.com/517546/122399400-0e227b80-cf7b-11eb-964a-f72611ec8fce.png)

Reply:
Maybe you didn't have the latest helm? That value should come from values.yaml

https://github.com/reanahub/reana/blob/2235424045d01483343e14cd90ea9e6cc292c37e/helm/reana/values.yaml#L62-L63

Reply:
I had, I tested with:

```console
$ reana-dev git-status -c ALL | grep -v '@ master'
docs.reana.io @ pr-84 @ f7cdef9 configuration: added configuring workflow scheduler page
reana-ui @ release-0.8.0-alpha.2 @ c9893b5 release: 0.8.0-alpha.2
reana-workflow-engine-yadage @ pr-194 @ 44ea0de release: 0.8.0a2
reana-message-broker @ pr-38 @ a869111 release: 0.8.0a1
reana @ pr-511 (1 AHEAD upstream/pr/511) @ b70c554 WIP
reana-server @ pr-377 @ a029855 global: refactor scheduling policy logic
```

and the only difference was REANA_MAX_CONCURRENT_BATCH_WORKFLOWS set to 1.

I'll rerun to confirm/infirm...

P.S. Perhaps we are not doing str->int cast along the way?

Reply:
I confirm I can reproduce the problem with the above setup. The first `run-ci` command pass fine, the three workflows are successful, but then when I submit several workflows with different memory limit requirements to test the queueing and scheduling policy, I am getting:

```
$ kubectl get pods | grep reana-server 
reana-server-694d7955ff-drhs6                                1/2     CrashLoopBackOff   2          12m
$ kubectl logs reana-server-694d7955ff-drhs6 scheduler | tail -5
  File "/usr/local/lib/python3.8/site-packages/kombu/messaging.py", line 590, in <listcomp>
    [callback(body, message) for callback in callbacks]
  File "/usr/local/lib/python3.8/site-packages/reana_server/scheduler.py", line 236, in on_message
    sleep(REANA_SCHEDULER_SECONDS_TO_WAIT_FOR_REANA_READY)
TypeError: an integer is required (got type str)
```

Reply:
> Perhaps we are not doing str->int cast along the way?

Indeed, in `reana_server/config.py` we are doing:

```python
REANA_SCHEDULER_SECONDS_TO_WAIT_FOR_REANA_READY = os.getenv(
    "REANA_SCHEDULER_SECONDS_TO_WAIT_FOR_REANA_READY"
)
```

which will produce a string not an integer... We need a (safe) cast there.